packages/next/next-server/server/image-optimizer.ts

```diff
@@ -136,17 +136,74 @@ export async function imageOptimizer(
     }
   }
 
-  const upstreamRes = await fetch(href)
+  let upstreamBuffer: Buffer
+  let upstreamType: any
+  let maxAge: any
+
+  if (absoluteUrl) {
+    const upstreamRes = await fetch(href)
+
+    if (!upstreamRes.ok) {
+      res.statusCode = upstreamRes.status
+      res.end('"url" parameter is valid but upstream response is invalid')
+      return { finished: true }
+    }
 
-  if (!upstreamRes.ok) {
     res.statusCode = upstreamRes.status
-    res.end('"url" parameter is valid but upstream response is invalid')
-    return { finished: true }
+    upstreamBuffer = Buffer.from(await upstreamRes.arrayBuffer())
+    upstreamType = upstreamRes.headers.get('Content-Type')
+    maxAge = getMaxAge(upstreamRes.headers.get('Cache-Control'))
+  } else {
+    try {
+      const _req: any = {
+        headers: req.headers,
+        method: req.method,
+        url: relativeUrl,
+      }
+      const resBuffers: Buffer[] = []
+      const mockRes: any = new Stream.Writable()
+
+      mockRes.write = (chunk: Buffer | string) => {
+        resBuffers.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk))
+      }
+      mockRes._write = (chunk: Buffer | string) => {
+        mockRes.write(chunk)
+      }
+
+      const mockHeaders: Record<string, string | string[]> = {}
+
+      mockRes.writeHead = (_status: any, _headers: any) =>
+        Object.assign(mockHeaders, _headers)
+      mockRes.getHeader = (name: string) => mockHeaders[name.toLowerCase()]
+      mockRes.getHeaders = () => mockHeaders
+      mockRes.getHeaderNames = () => Object.keys(mockHeaders)
+      mockRes.setHeader = (name: string, value: string | string[]) =>
+        (mockHeaders[name.toLowerCase()] = value)
+      mockRes._implicitHeader = () => {}
+      mockRes.finished = false
+      mockRes.statusCode = 200
+
+      await server.getRequestHandler()(
+        _req,
+        mockRes,
+        nodeUrl.parse(relativeUrl!, true)
+      )
+      res.statusCode = mockRes.statusCode
+
+      // make sure to 404 for non-static file requests
+      if (!mockRes.servedStatic) {
+        throw new Error('non-static file requested')
```

Comment:
Should this `return` instead of `throw`?

Reply:
Should we return 404 here instead? I have it re-using the return in the catch below currently

Reply:
I think so? I was reading the comment that says 404 but the actual response is 500. I'm not sure why we wouldn't allow non-static though 🤔

Reply:
Updated as discussed to allow API endpoints to potentially return image content